### PR TITLE
Remove extraneous nil checks before d.Set()

### LINKED ISSUE #15893: provider: Remove extraneous nil checks before d.Set()
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The [`(*schema.ResourceData).Set()` receiver method](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#ResourceData.Set) automatically handles pointer types and `nil` values, so checking for `nil` beforehand is extraneous.

### Example semgrep Rule

```yaml
rules:
  - id: helper-schema-ResourceData-Set-extraneous-nil-check
    languages: [go]
    message: Nil value check before `d.Set()` is extraneous
    paths:
      include:
        - aws/
    patterns:
      - pattern-either:
        - pattern: |
            if $APIOBJECT != nil {
              d.Set($ATTRIBUTE, $APIOBJECT)
            }
        - pattern: |
            if $APIOBJECT != nil {
              if $ERROR := d.Set($ATTRIBUTE, $APIOBJECT); err != nil { $BODY }
            }
        - pattern: |
            if $APIOBJECT != nil {
              ...
              $ERROR := d.Set($ATTRIBUTE, $APIOBJECT)
              ...
            }
        - pattern: |
            if $APIOBJECT != nil {
              ...
              $ERROR = d.Set($ATTRIBUTE, $APIOBJECT)
              ...
            }
    severity: WARNING

```

### Example Reports

```
aws/cloudfront_distribution_configuration_structure.go
severity:warning rule:helper-schema-ResourceData-Set-extraneous-nil-check: Nil value check before `d.Set()` is extraneous
98:	if distributionConfig.CallerReference != nil {
99:		d.Set("caller_reference", distributionConfig.CallerReference)
100:	}
106:	if distributionConfig.DefaultRootObject != nil {
107:		d.Set("default_root_object", distributionConfig.DefaultRootObject)
108:	}
109:	if distributionConfig.HttpVersion != nil {
110:		d.Set("http_version", distributionConfig.HttpVersion)
111:	}
112:	if distributionConfig.WebACLId != nil {
113:		d.Set("web_acl_id", distributionConfig.WebACLId)
114:	}

aws/data_source_aws_ami.go
severity:warning rule:helper-schema-ResourceData-Set-extraneous-nil-check: Nil value check before `d.Set()` is extraneous
258:	if image.Description != nil {
259:		d.Set("description", image.Description)
260:	}
264:	if image.ImageOwnerAlias != nil {
265:		d.Set("image_owner_alias", image.ImageOwnerAlias)
266:	}
268:	if image.KernelId != nil {
269:		d.Set("kernel_id", image.KernelId)
270:	}
273:	if image.Platform != nil {
274:		d.Set("platform", image.Platform)
275:	}
277:	if image.RamdiskId != nil {
278:		d.Set("ramdisk_id", image.RamdiskId)
279:	}
280:	if image.RootDeviceName != nil {
281:		d.Set("root_device_name", image.RootDeviceName)
282:	}
285:	if image.SriovNetSupport != nil {
286:		d.Set("sriov_net_support", image.SriovNetSupport)
287:	}

aws/data_source_aws_elasticache_cluster.go
severity:warning rule:helper-schema-ResourceData-Set-extraneous-nil-check: Nil value check before `d.Set()` is extraneous
192:	if cluster.ReplicationGroupId != nil {
193:		d.Set("replication_group_id", cluster.ReplicationGroupId)
194:	}
```

### Affected Resources

TBD

### Definition of Done

* Rule added to `.semgrep.yml` with no reports on main branch and CI passes

### PR DESCRIPTION
**Pull request by mattburgess**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15893

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
NONE
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: aws/data_source_aws_rds_cluster.go [hunk 1]
```diff
@@ -221,7 +221,7 @@ func dataSourceAwsRdsClusterRead(d *schema.ResourceData, meta interface{}) error
 	// RDS accepts a DatabaseName but does not return it, causing a perpetual
 	// diff.
 	//	See https://github.com/hashicorp/terraform/issues/4671 for backstory
-	if dbc.DatabaseName != nil {
+	if dbc.DatabaseName != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/data_source_aws_rds_cluster.go:224]: I do wonder if the API bug discussed here has been fixed yet; but for now the safest option is probably just to leave this well alone!

File: aws/data_source_aws_s3_bucket_object.go [hunk 2]
```diff
@@ -185,7 +185,7 @@ func dataSourceAwsS3BucketObjectRead(d *schema.ResourceData, meta interface{}) e
 	// The "STANDARD" (which is also the default) storage
 	// class when set would not be included in the results.
 	d.Set("storage_class", s3.StorageClassStandard)
-	if out.StorageClass != nil {
+	if out.StorageClass != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/data_source_aws_s3_bucket_object.go:188]: I considered reversing the check here, i.e.:

```
if out.StorageClass == nil {
    d.Set("storage_class", s3.StorageClassStandard)
} else {
    d.Set("storage_class", out.StorageClass)
}
```
Personally, I find that makes the intent clearer but ignoring the lint also works for me.

File: aws/resource_aws_api_gateway_authorizer.go [hunk 3]
```diff
@@ -152,7 +152,7 @@ func resourceAwsApiGatewayAuthorizerRead(d *schema.ResourceData, meta interface{
 
 	d.Set("authorizer_credentials", authorizer.AuthorizerCredentials)
 
-	if authorizer.AuthorizerResultTtlInSeconds != nil {
+	if authorizer.AuthorizerResultTtlInSeconds != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/resource_aws_api_gateway_authorizer.go:155]: Again, reversing the test here might make intent clearer?

File: aws/resource_aws_eip.go [hunk 4]
```diff
@@ -260,12 +260,12 @@ func resourceAwsEipRead(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	d.Set("association_id", address.AssociationId)
-	if address.InstanceId != nil {
+	if address.InstanceId != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/resource_aws_eip.go:263]: Reverse these tests too?

File: aws/resource_aws_rds_cluster.go [hunk 5]
```diff
@@ -1033,7 +1033,7 @@ func resourceAwsRDSClusterRead(d *schema.ResourceData, meta interface{}) error {
 	// RDS accepts a DatabaseName but does not return it, causing a perpetual
 	// diff.
 	//	See https://github.com/hashicorp/terraform/issues/4671 for backstory
-	if dbc.DatabaseName != nil {
+	if dbc.DatabaseName != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/resource_aws_rds_cluster.go:1036]: Same API bug as above.

File: aws/resource_aws_s3_bucket_object.go [hunk 6]
```diff
@@ -366,7 +366,7 @@ func resourceAwsS3BucketObjectRead(d *schema.ResourceData, meta interface{}) err
 	// The "STANDARD" (which is also the default) storage
 	// class when set would not be included in the results.
 	d.Set("storage_class", s3.StorageClassStandard)
-	if resp.StorageClass != nil {
+	if resp.StorageClass != nil { // nosem: helper-schema-ResourceData-Set-extraneous-nil-check
```
Comment (mattburgess) [inline aws/resource_aws_s3_bucket_object.go:369]: And this is the same logic as its datasource counterpart, so checks could be reversed.

File: aws/resource_aws_eip.go (hunk 4 above)
Comment (ewbankkit) [reply aws/resource_aws_eip.go:271]: In fact setting the resource data field to `""` is what happens if the passed in value is `nil`, so the `else` branch is not needed.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

**Comment by ewbankkit** [comment]
@mattburgess Thanks for the contribution :tada: :clap:.

**Comment by mattburgess** [comment]
Oops - sorry @ewbankkit but the semgrep rule wasn't correct (it was checking `aws/` not `internal/`. I'll fix that up and the resultant semgrep failures in another branch. I really should have set this back to draft until I was confident I'd caught everything!

File: internal/service/apigateway/integration.go [hunk 7]
```diff
@@ -274,9 +274,7 @@ func resourceIntegrationRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("cache_namespace", integration.CacheNamespace)
 	d.Set("connection_id", integration.ConnectionId)
 	d.Set("connection_type", apigateway.ConnectionTypeInternet)
-	if integration.ConnectionType != nil {
-		d.Set("connection_type", integration.ConnectionType)
-	}
+	d.Set("connection_type", integration.ConnectionType)
```
Comment (jufemaiz) [inline internal/service/apigateway/integration.go:277]: This is not right - the nil check is needed here.

Issues are arising due to this. See: https://github.com/hashicorp/terraform-provider-aws/issues/28997

Comment (ewbankkit) [reply internal/service/apigateway/integration.go:277]: Fix in https://github.com/hashicorp/terraform-provider-aws/pull/29016.

**Pull request merged**